### PR TITLE
Broadcast binary ops with device scalars

### LINKED ISSUE #498: Broadcasting with device scalars?
Raised by @kaushikcfd here: https://github.com/inducer/arraycontext/issues/49#issuecomment-869266944

Example:
```
>>> a = cla.to_device(cq, np.ones(3))
>>> b = cla.to_device(cq, np.ones(()))
>>> a
cl.Array([1., 1., 1.])
>>> b
cl.Array(1.)
>>> a * b
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "/shared/home/kgk2/pack/virtual_envs/emirge/pyopencl/pyopencl/array.py", line 1161, in __mul__
    self._elwise_multiply(result, self, other))
  File "/shared/home/kgk2/pack/virtual_envs/emirge/pyopencl/pyopencl/array.py", line 183, in kernel_runner
    knl = kernel_getter(*args, **kwargs)
  File "/shared/home/kgk2/pack/virtual_envs/emirge/pyopencl/pyopencl/array.py", line 897, in _elwise_multiply
    assert out.shape == b.shape
```

### PR DESCRIPTION
**Pull request by kaushikcfd**
- Closes #498

### CODE REVIEW

File: pyopencl/array.py [hunk 1]
```diff
@@ -874,11 +892,13 @@ def __hash__(self):
     def _axpbyz(out, afac, a, bfac, b, queue=None):
         """Compute ``out = selffac * self + otherfac*other``,
         where *other* is an array."""
-        assert out.shape == a.shape
-        assert out.shape == b.shape
-
+        assert (a.shape == b.shape
+                or a.shape == ()
+                or b.shape == ())
```
Comment (inducer) [inline pyopencl/array.py]: This loses the checking of `out.shape`.

File: pyopencl/array.py [hunk 2]
```diff
@@ -893,10 +913,14 @@ def _axpbz(out, a, x, b, queue=None):
     @staticmethod
     @elwise_kernel_runner
     def _elwise_multiply(out, a, b, queue=None):
-        assert out.shape == a.shape
-        assert out.shape == b.shape
+        assert (a.shape == b.shape
+                or a.shape == ()
+                or b.shape == ())
```
Comment (inducer) [inline pyopencl/array.py]: This loses the checking of `out.shape`.

File: pyopencl/array.py [hunk 3]
```diff
@@ -910,11 +934,14 @@ def _rdiv_scalar(out, ary, other, queue=None):
     @elwise_kernel_runner
     def _div(out, self, other, queue=None):
         """Divides an array by another array."""
-
-        assert self.shape == other.shape
+        assert (other.shape == self.shape
+                or other.shape == ()
+                or self.shape == ())
```
Comment (inducer) [inline pyopencl/array.py]: Check `out`?

File: pyopencl/array.py [hunk 4]
```diff
@@ -1027,10 +1054,13 @@ def _scalar_binop(out, a, b, queue=None, op=None):
     @staticmethod
     @elwise_kernel_runner
     def _array_binop(out, a, b, queue=None, op=None):
-        if a.shape != b.shape:
-            raise ValueError("shapes of binop arguments do not match")
+        assert (a.shape == b.shape
+                or a.shape == ()
+                or b.shape == ())
```
Comment (inducer) [inline pyopencl/array.py]: Check `out`?

File: pyopencl/array.py [hunk 5]
```diff
@@ -104,6 +104,24 @@ def _get_truedivide_dtype(obj1, obj2, queue):
...
+        return obj1._new_like_me(dtype_getter(obj1, obj2, cq),
+                                 cq)
+    elif obj1.shape == ():
+        return obj2._new_like_me(dtype_getter(obj1, obj2, cq),
+                                 cq)
+    elif obj2.shape == ():
+        return obj1._new_like_me(dtype_getter(obj1, obj2, cq),
+                                 cq)
+    else:
+        raise NotImplementedError("Broadcasting binary op with shapes:"
```
Comment (inducer) [inline pyopencl/array.py]:
```suggestion
        raise NotImplementedError("Broadcasting binary operator with shapes:"
```

File: pyopencl/array.py [hunk 6]
```diff
@@ -104,6 +104,24 @@ def _get_truedivide_dtype(obj1, obj2, queue):
     return result
 
 
+def _get_broadcasted_binary_op_result(obj1, obj2, cq, dtype_getter=None):
+    if dtype_getter is None:
+        dtype_getter = _get_common_dtype
```
Comment (inducer) [inline pyopencl/array.py]:
```suggestion
def _get_broadcasted_binary_op_result(obj1, obj2, cq, dtype_getter=_get_common_dtype):
```

File: pyopencl/array.py [hunk 7]
```diff
@@ -104,6 +104,24 @@ def _get_truedivide_dtype(obj1, obj2, queue):
...
+
+    if obj1.shape == obj2.shape:
+        return obj1._new_like_me(dtype_getter(obj1, obj2, cq),
+                                 cq)
+    elif obj1.shape == ():
+        return obj2._new_like_me(dtype_getter(obj1, obj2, cq),
+                                 cq)
+    elif obj2.shape == ():
+        return obj1._new_like_me(dtype_getter(obj1, obj2, cq),
```
Comment (inducer) [inline pyopencl/array.py:117]: Doesn't this subsume the `assert`s below?

**Review by inducer — COMMENTED** [review]
Thanks! Just a few minor things below, then this is ready to go in.

**Comment by inducer** [comment]
Unsubscribing... @-mention or request review once it's ready for a look or needs attention.

File: pyopencl/array.py (hunk 6 above)
Comment (kaushikcfd) [reply pyopencl/array.py]: Thanks, 698423f

File: pyopencl/array.py (hunk 4 above)
Comment (kaushikcfd) [reply pyopencl/array.py]: https://github.com/inducer/pyopencl/commit/9ee35f26a7881655d6b715afe55013bef533ebd2

File: pyopencl/array.py (hunk 3 above)
Comment (kaushikcfd) [reply pyopencl/array.py]: https://github.com/inducer/pyopencl/commit/9ee35f26a7881655d6b715afe55013bef533ebd2

File: pyopencl/array.py (hunk 2 above)
Comment (kaushikcfd) [reply pyopencl/array.py]: https://github.com/inducer/pyopencl/commit/9ee35f26a7881655d6b715afe55013bef533ebd2

File: pyopencl/array.py (hunk 1 above)
Comment (kaushikcfd) [reply pyopencl/array.py]: https://github.com/inducer/pyopencl/commit/9ee35f26a7881655d6b715afe55013bef533ebd2

File: pyopencl/array.py (hunk 7 above)
Comment (kaushikcfd) [reply pyopencl/array.py:117]: Added it as there isn't quite a promise that control reaching those asserts would have reached "_get_broadcasted_binary_op_result".

**Comment by inducer** [comment]
Thanks!

**Comment by inducer** [comment]
We probably want a release with that. I'll do that tomorrow, assuming this merges OK.

**Comment by inducer** [comment]
https://pypi.org/project/pyopencl/2021.2.6/

**Pull request merged**